sysl2/sysl/validate.go

```diff
@@ -0,0 +1,377 @@
+			MsgType: ERROR,
+		})
+	}
+
+	assignCount := 0
+	for _, stmt := range view.GetExpr().GetTransform().GetStmt() {
+		if stmt.GetAssign() != nil {
+			if assignCount == 0 && stmt.GetAssign().GetName() != viewName {
+				validationMsgs = append(validationMsgs, ValidationMsg{
+					Message: "In view 'filename' : missing type: 'filename'",
+					MsgType: ERROR,
+				})
+			} else if assignCount > 0 {
+				validationMsgs = append(validationMsgs, ValidationMsg{
+					Message: fmt.Sprintf("In view 'filename' : Excess assignments: '%s'", stmt.GetAssign().GetName()),
+					MsgType: ERROR,
+				})
+			}
+			assignCount++
+		}
+	}
+	return validationMsgs
+}
+
+func getImplAttrNames(attrs map[string]*sysl.Type) map[string]struct{} {
+	implAttrNames := map[string]struct{}{}
+
+	for attrName := range attrs {
+		implAttrNames[attrName] = struct{}{}
+	}
+
+	return implAttrNames
+}
+
+func compareSpecVsImpl(grammarSpec, specAttrs, implAttrs map[string]*sysl.Type,
+	implAttrNames map[string]struct{}, viewName string) []ValidationMsg {
+	validationMsgs := make([]ValidationMsg, 0, len(implAttrs))
+
+	for gk, gv := range specAttrs {
+		if grammarSpec[gk].GetOneOf() != nil {
+			validationMsgs = append(validationMsgs, compareOneOf(grammarSpec, implAttrs, implAttrNames, viewName, gk)...)
+		} else if _, exists := implAttrs[gk]; !exists {
+			if !gv.GetOpt() {
+				validationMsgs = append(validationMsgs, ValidationMsg{
+					Message: fmt.Sprintf("In view '%s', type '%s' is missing", viewName, gk),
+					MsgType: ERROR,
+				})
+			}
+		} else {
+			delete(implAttrNames, gk)
+		}
+	}
+
+	for attrName := range implAttrNames {
+		validationMsgs = append(validationMsgs, ValidationMsg{
+			Message: fmt.Sprintf("In view '%s', excess attribute is defined: '%s'", viewName, attrName),
+			MsgType: ERROR,
+		})
+		delete(implAttrNames, attrName)
+	}
+
+	return validationMsgs
+}
+
+func compareOneOf(grammarSpec, implAttrs map[string]*sysl.Type,
+	implAttrNames map[string]struct{}, viewName, typeName string) []ValidationMsg {
+	validationMsgs := make([]ValidationMsg, 0, len(implAttrs))
+	matching := true
+
+	for _, one := range grammarSpec[typeName].GetOneOf().GetType() {
+		name := one.GetTypeRef().GetRef().GetPath()[0]
+
+		if strings.Index(name, "__Choice_Combination_") == 0 {
+			if len(implAttrs) == 1 {
+				continue
+			}
+
+			validationMsgs = append(validationMsgs, compareSpecVsImpl(grammarSpec,
+				grammarSpec[name].GetTuple().GetAttrDefs(), implAttrs, implAttrNames, viewName)...)
+			break
+		} else {
+			if _, exists := implAttrs[name]; !exists {
+				matching = false
+			} else {
+				matching = true
+				delete(implAttrNames, name)
+				break
+			}
+		}
+	}
+
+	if !matching {
+		implAttrName := ""
+		for k := range implAttrs {
+			implAttrName = k
+		}
+		validationMsgs = append(validationMsgs, ValidationMsg{
+			Message: fmt.Sprintf("In view '%s', invalid choice has been made as : '%s'", viewName, implAttrName),
```

Comment:
`In View %s, (%s=attrName) does not match any of the options for return Type (%s)`